File: bindings/android/SpektreSurface.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
//
// SpektreSurface.kt — Kotlin façade over v76 σ-Surface on Android.
//
// Design goals
// ============
//   * Zero Android framework dependency in the library.  Callers
//     adapt MotionEvent → Touch themselves (a 5-line mapping).
//   * Safe `@JvmStatic` entry points so Java code can call without
//     Kotlin interop.
//   * Every API returns a deterministic verdict from the C kernel;
//     no JVM allocation on the hot path once the JNI library is
//     loaded.
//
// Usage:
//   val v = SpektreSurface.version
//   val (hv, ok) = SpektreSurface.decode(Touch(0.5, 0.5, 0.8, phase=1, timestampMs=System.currentTimeMillis()))
//   val allow = SpektreSurface.decide(bits16)

package dev.spektre.surface

object SpektreSurface {
    init { System.loadLibrary("spektre_surface") }

    data class Touch(
        val xQ15: Int,
        val yQ15: Int,
        val pressureQ15: Int,
        val phase: Int,
        val timestampMs: Long,
    ) {
        constructor(x: Double, y: Double, pressure: Double,
                    phase: Int, timestampMs: Long) :
            this(
                xQ15        = (x.coerceIn(0.0, 1.0)        * 32767.0).toInt(),
                yQ15        = (y.coerceIn(0.0, 1.0)        * 32767.0).toInt(),
                pressureQ15 = (pressure.coerceIn(0.0, 1.0) * 32767.0).toInt(),
                phase       = phase,
                timestampMs = timestampMs,
            )
    }

    data class Hv(val w0: Long, val w1: Long, val w2: Long, val w3: Long)

    val version: String get() = nativeVersion()

    fun decode(t: Touch): Pair<Hv, Boolean> {
        val words = nativeTouchDecode(t.xQ15, t.yQ15, t.pressureQ15, t.phase, t.timestampMs)
        val ok = nativeTouchOk(t.xQ15, t.yQ15, t.pressureQ15, t.phase, t.timestampMs)
        return Hv(words[0], words[1], words[2], words[3]) to ok
    }

    /**
     * Composed 16-bit decision.  [bits] must contain exactly 16 elements,
     * each 0 or 1, in kernel order: v60, v61, v62, v63, v64, v65, v66,
     * v67, v68, v69, v70, v71, v72, v73, v74, v76.
     */
    fun decide(bits: ByteArray): Boolean {
        require(bits.size == 16) { "decide requires 16 bits" }
        return nativeDecide(bits)
    }

    @JvmStatic external fun nativeVersion(): String
    @JvmStatic external fun nativeTouchOk(
        xQ15: Int, yQ15: Int, pressureQ15: Int, phase: Int, timestampMs: Long): Boolean
    @JvmStatic external fun nativeTouchDecode(
        xQ15: Int, yQ15: Int, pressureQ15: Int, phase: Int, timestampMs: Long): LongArray
    @JvmStatic external fun nativeDecide(bits: ByteArray): Boolean
}

File: bindings/ios/SpektreSurface.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
//
// SpektreSurface.swift — thin Swift façade over v76 σ-Surface C ABI.
//
// Design goals
// ============
//   * Zero allocation on the hot path.  The C kernel already uses
//     aligned_alloc(64) internally; Swift just borrows pointers.
//   * UITouch / UIEvent interop without UIKit leakage.  We expose a
//     pure-Swift `Touch` struct that the caller is responsible for
//     populating from UITouch (a trivial 5-assignment mapping).
//   * Apple-tier usability.  Every call site reads like idiomatic
//     Swift; every verdict returns `true` iff the 16-kernel stack
//     allowed the action.
//   * WhatsApp / Telegram / Signal / iMessage / RCS / Matrix / XMPP /
//     Discord / Slack / Line ids mirror the C enum order so switch
//     statements stay exhaustive.
//
// Nothing here is thread-unsafe — the C kernel is pure + branchless
// and all state pointers are owned by the caller.

import Foundation

public enum Spektre {

    public struct Touch: Sendable {
        public var xQ15: UInt16
        public var yQ15: UInt16
        public var pressureQ15: UInt16
        public var phase: UInt16      // 0 began, 1 moved, 2 ended, 3 cancelled
        public var timestampMs: UInt32

        public init(xQ15: UInt16, yQ15: UInt16, pressureQ15: UInt16,
                    phase: UInt16, timestampMs: UInt32) {
            self.xQ15 = xQ15; self.yQ15 = yQ15
            self.pressureQ15 = pressureQ15; self.phase = phase
            self.timestampMs = timestampMs
        }

        /// Convenience from UITouch-style normalised (0..1) floats.
        public init(x: Double, y: Double, pressure: Double,
                    phase: UInt16, timestampMs: UInt32) {
            func q(_ v: Double) -> UInt16 {
                let s = max(0.0, min(1.0, v)) * 32767.0
                return UInt16(s.rounded())
            }
            self.init(xQ15: q(x), yQ15: q(y), pressureQ15: q(pressure),
                      phase: phase, timestampMs: timestampMs)
        }
    }

    public enum Protocol_: UInt8, Sendable, CaseIterable {
        case whatsApp = 0, telegram, signal, iMessage, rcs
        case matrix, xmpp, discord, slack, line
    }

    public enum Gesture: UInt32, Sendable, CaseIterable {
        case tap = 0, doubleTap, longPress, swipe, pinch, rotate
    }

    // MARK: - Touch

    public static func decode(_ t: Touch) -> (hv: (UInt64, UInt64, UInt64, UInt64), ok: Bool) {
        var raw = cos_v76_touch_t(x_q15: t.xQ15, y_q15: t.yQ15,
                                  pressure_q15: t.pressureQ15,
                                  phase: t.phase,
                                  timestamp_ms: t.timestampMs)
        var hv = cos_v76_hv_t(w: (0, 0, 0, 0))
        cos_v76_touch_decode(&hv, &raw)
        let ok = cos_v76_touch_ok(&raw) == 1
        return ((hv.w.0, hv.w.1, hv.w.2, hv.w.3), ok)
    }

    // MARK: - Composed decision (16-kernel)

    /// Returns true iff every one of the 16 kernels allows.
    public static func decide(_ bits: [UInt8]) -> Bool {
        precondition(bits.count == 16, "Spektre.decide requires 16 bits")
        let d = cos_v76_compose_decision(
            bits[0],  bits[1],  bits[2],  bits[3],
            bits[4],  bits[5],  bits[6],  bits[7],
            bits[8],  bits[9],  bits[10], bits[11],
            bits[12], bits[13], bits[14], bits[15])
        return d.allow == 1
    }

    // MARK: - Version

    public static var version: String {
        String(cString: cos_v76_version)
    }
}
